experimental/AlgebraicStatistics/src/Markov.jl

```diff
@@ -0,0 +1,331 @@
+end
+
+function state_space_indices(R::MarkovRing, K=R.random_variables)
+  idx = find_random_variables(R, K)
+  return length(idx) == 1 ? range(1, length(R.state_spaces[idx[1]])) : Iterators.product([range(1, length(R.state_spaces[i])) for i in idx]...)
+end
+
+# -*- CI equations for MarkovRing -*-
+
+# p is a permutation of 1:n and x is a vector of length n.
+# This method returns the components of x permuted by p,
+# e.g. apply_permutation([3,2,1], [0,1,2]) == [2,1,0].
+function apply_permutation(p, x)
+  px = Vector{Int64}(undef, length(p))
+  for i in p
+    px[p[i]] = x[i]
+  end
+  return px
+end
+
+@doc raw"""
+    marginal(R::MarkovRing, K, x)
+
+Return a marginal as a sum of unknowns from `R`. The argument `K` lists
+random variables which are fixed to the event `x`; all other random
+variables in `R` are summed over their respective state spaces.
+
+## Examples
+
+``` jldoctest
+julia> R = markov_ring("A" => 1:2, "B" => 1:2, "X" => 1:2, "Y" => 1:2)
+MarkovRing for random variables A -> {1, 2}, B -> {1, 2}, X -> {1, 2}, Y -> {1, 2} in 16 variables over Rational field
+
+julia> marginal(R, ["A", "X"], [1,2])
+p[1, 1, 2, 1] + p[1, 2, 2, 1] + p[1, 1, 2, 2] + p[1, 2, 2, 2]
+```
+"""
+function marginal(R::MarkovRing, K, x)
+  p = unknowns(R)
+  N = random_variables(R)
+  M = setdiff(N, K)
+  summands = Vector()
+  for y in state_space(R, M)
+    idx = apply_permutation(
+      find_random_variables(R, [K..., M...]),
+      [x..., y...]
+    )
+    push!(summands, p[idx...])
+  end
+  return length(summands) == 1 ? summands[1] : +(summands...)
+end
+
+@doc raw"""
+    ci_ideal(R::MarkovRing, stmts)::MPolyIdeal
+
+Return the ideal for the conditional independence statements
+given by `stmts`.
+
+## Examples
+
+``` jldoctest
+julia> R = markov_ring("A" => 1:2, "B" => 1:2, "X" => 1:2)
+MarkovRing for random variables A -> {1, 2}, B -> {1, 2}, X -> {1, 2} in 8 variables over Rational field
+
+julia> ci_ideal(R, [CI"X,A|B", CI"X,B|A"])
+Ideal generated by
+  p[1, 1, 1]*p[2, 1, 2] - p[2, 1, 1]*p[1, 1, 2]
+  p[1, 2, 1]*p[2, 2, 2] - p[2, 2, 1]*p[1, 2, 2]
+  p[1, 1, 1]*p[1, 2, 2] - p[1, 2, 1]*p[1, 1, 2]
+  p[2, 1, 1]*p[2, 2, 2] - p[2, 2, 1]*p[2, 1, 2]
+```
+"""
+function ci_ideal(R::MarkovRing, stmts)::MPolyIdeal
+  eqs = Vector()
+  for stmt in stmts
+    # The proper CI equations are the familiar 2x2 determinants in
+    # marginals of the probability tensor.
+    # TODO: Functional dependence not yet supported.
+    QI = state_space(R, stmt.I)
+    QJ = state_space(R, stmt.J)
+    IJK = [stmt.I..., stmt.J..., stmt.K...]
+    for z in state_space(R, stmt.K)
+      for x in subsets(collect(QI), 2)
+        for y in subsets(collect(QJ), 2)
+          p11 = marginal(R, IJK, [x[1], y[1], z...]);
+          p12 = marginal(R, IJK, [x[1], y[2], z...]);
+          p21 = marginal(R, IJK, [x[2], y[1], z...]);
+          p22 = marginal(R, IJK, [x[2], y[2], z...]);
+          push!(eqs, p11*p22 - p12*p21)
+        end
+      end
+    end
```

Comment:
```suggestion
    for z in state_space(R, stmt.K), x in subsets(collect(QI), 2), y in subsets(collect(QJ), 2)
        p11 = marginal(R, IJK, [x[1], y[1], z...]);
        p12 = marginal(R, IJK, [x[1], y[2], z...]);
        p21 = marginal(R, IJK, [x[2], y[1], z...]);
        p22 = marginal(R, IJK, [x[2], y[2], z...]);
        push!(eqs, p11*p22 - p12*p21)
    end
```
The loops seem not to depend on each other, right?